### PR TITLE
Added new api call plus other changes (v2)

### PR DESCRIPTION
**Pull request by loco-choco**
* Made it [join translation tables from multiple mods for the same languages](https://github.com/ShoosGun/outer-wilds-localization-utility/blob/457a2df84809aa082bdc538a420267fe51df78a0/LocalizationUtility/Patches/TextTranslationPatches.cs#L33), being them custom or vanilla.

* Added the api call [`AddTranslation`](https://github.com/ShoosGun/outer-wilds-localization-utility/blob/c1eeda080c1f7aab7e2c897d409208073d59c743/LocalizationUtility/LocalizationAPI.cs#L12).

* Made the mod "see" the vanilla languages as ["custom languages"](https://github.com/ShoosGun/outer-wilds-localization-utility/blob/c1eeda080c1f7aab7e2c897d409208073d59c743/LocalizationUtility/LocalizationUtility.cs#L176).

* Changed how it loads translation tables by first [caching them in a dictionary](https://github.com/ShoosGun/outer-wilds-localization-utility/blob/457a2df84809aa082bdc538a420267fe51df78a0/LocalizationUtility/Patches/TextTranslationPatches.cs#L15), and then [appending/editing the game's translation table on the prefix/postfix](https://github.com/ShoosGun/outer-wilds-localization-utility/blob/457a2df84809aa082bdc538a420267fe51df78a0/LocalizationUtility/Patches/TextTranslationPatches.cs#L102).

* Made other changes I can't really remember.

### CODE REVIEW

**Comment by loco-choco** [comment]
This time I tested with just Czech Localization, just Celeste Wilds, and both, and it all worked as expected.

**Comment by loco-choco** [comment]
Here is the example in [Celeste Wilds of the new api call](https://github.com/ShoosGun/CelesteWilds/blob/fb06df9a1cad5a1e7aa8d7894f378fdeffb7371c/ModTemplate/Utils.cs#L45) (that was on the now closed pr).

**Comment by loco-choco** [comment]
Added with this last commit 8 more API calls:
Add entries directly to the regular translation table.
```CSharp
        public void AddRegularTranslation(string name, string key, string value);
        public void AddRegularTranslation(string name, string commonKeyPrefix, params string[] entries);
        public void AddRegularTranslation(string name, params KeyValuePair<string, string>[] entries);
```
Add entries directly to the shiplog translation table.
```CSharp
        public void AddShiplogTranslation(string name, string key, string value);
        public void AddShiplogTranslation(string name, string commonKeyPrefix, params string[] entries);
        public void AddShiplogTranslation(string name, params KeyValuePair<string, string>[] entries);
```
Add entries directly to the ui translation table.
```CSharp
       public void AddUITranslation(string name, int key, string value);
        public void AddUITranslation(string name, params KeyValuePair<int, string>[] entries);
```

`AddRegularTranslation(string name, string commonKeyPrefix, params string[] entries)` and `AddShiplogTranslation(string name, string commonKeyPrefix, params string[] entries)` will automatically add the prefix you choose that most of the text have as their keys. So if you have a sign text, you can pass the text as the entries, and for the `commonKeyPrefix` just need to pass `Sign`.

**Comment by loco-choco** [comment]
Added another api call:
```CSharp
void AddTranslation(string name, KeyValuePair<string, string>[] regularEntries, KeyValuePair<string, string>[] shipLogEntries, KeyValuePair<int, string>[] uiEntries);        
```
It simmilar to calling `AddRegularTranslation`, `AddShiplogTranslation` and `AddUITranslation` one after the other. The idea behind it is that, if you are using another format to store the translations (like a json), you only need to find a way to get the entries' key and value for the utility to add them to the correct tables.

File: LocalizationUtility/ILocalizationAPI.cs [hunk 1]
```diff
@@ -1,12 +1,23 @@
...
 using System;
+using System.Collections.Generic;
 
 namespace LocalizationUtility
 {
     public interface ILocalizationAPI
     {
         void RegisterLanguage(ModBehaviour mod, string name, string translationPath);
         void RegisterLanguage(ModBehaviour mod, string name, string translationPath, string languageToReplace);
+        void AddTranslation(ModBehaviour mod, string name, string translationPath);
```
Comment (xen-42) [inline LocalizationUtility/ILocalizationAPI.cs]: Should clarify that `name` is `languageName`

File: LocalizationUtility/LocalizationUtility.cs [hunk 2]
```diff
@@ -66,15 +89,123 @@ public void RegisterLanguage(ModBehaviour mod, string name, string translationPa
...
+                }
+                WriteError($"The custom language {name} isn't registered yet");
+
+            }
+            catch (Exception ex)
+            {
+                WriteError($"Failed to add translation to language {name}. {ex}");
+            }
+        }
+        public void AddRegularTranslation(string name, string commonKeyPrefix, params string[] entries)
```
Comment (xen-42) [inline LocalizationUtility/LocalizationUtility.cs]: Why do we need two different versions of each method here?

File: LocalizationUtility/Patches/TextTranslationPatches.cs [hunk 3]
```diff
@@ -80,26 +69,228 @@ public static bool TextTranslation_SetLanguage(ref TextTranslation.Language lang
...
+            var onLanguageChanged = (MulticastDelegate)__instance.GetType().GetField("OnLanguageChanged", BindingFlags.Instance | BindingFlags.NonPublic).GetValue(__instance);
+            if (onLanguageChanged != null)
+            {
+                onLanguageChanged.DynamicInvoke();
+            }
             return false;
         }
 
+        //Maybe call this when OnLanguageChanged gets called, but could cause issues, I really don't know
+        //A *transpiler* could help to call this right after the xml is loaded (Xen would kill me)
```
Comment (xen-42) [inline LocalizationUtility/Patches/TextTranslationPatches.cs]: Why would I kill you tho

**Review by xen-42 — CHANGES_REQUESTED** [review]
Can you merge dev into this there's some conflicts that I don't get.

Do any of the patches conflict with the current NH or QSB translation systems? Mostly concerned about QSB because I don't see them supporting this.

File: LocalizationUtility/LocalizationUtility.cs (hunk 2 above)
Comment (loco-choco) [reply LocalizationUtility/LocalizationUtility.cs]: The idea is that you should use `public void AddRegularTranslation(string name, string commonKeyPrefix, params string[] entries)` only if they have the same prefix for the key of all the entries, like the translation entries from sign texts. And `public void AddRegularTranslation(string languageName, params KeyValuePair<string, string>[] entries)` if you are getting the translations from somewhere where the key prefix can be different for each entry, so loading from files is easier. It is supposed to work as a version of `public void AddTranslation(string languageName, KeyValuePair<string, string>[] regularEntries, KeyValuePair<string, string>[] shipLogEntries, KeyValuePair<int, string>[] uiEntries)` but where you only care about the regular translation table.

**Comment by loco-choco** [comment]
About the possible conflict with NH or QSB. I fell that there is almost no way for it to conflict with the postfix, as it just adds to the current translation table, nothing else. The prefix might cause issues, as it skips the game code when it is a custom language. I'll check both patches of nh and qsb to search for issues.

Edit: Perhaps instead of skipping but changing the xml file with a transpiler might help.

File: LocalizationUtility/Patches/TextTranslationPatches.cs (hunk 3 above)
Comment (loco-choco) [reply LocalizationUtility/Patches/TextTranslationPatches.cs]: Does this mean I'm allowed to make a transpiler to do that? (usually people don't like them)

**Comment by loco-choco** [comment]
Aight, now it should be good to merge without conflicts

File: LocalizationUtility/Patches/TextTranslationPatches.cs (hunk 3 above)
Comment (xen-42) [reply LocalizationUtility/Patches/TextTranslationPatches.cs]: idk why don't they like them? If it works without one thats fine then right?

Comment (loco-choco) [reply LocalizationUtility/Patches/TextTranslationPatches.cs]: I guess so, well I changed the system to work with a transpiler. I think this way it is more compatible with nh and qsb as we aren't not running the original method.

**Comment by loco-choco** [comment]
As I mentioned on the comment , this last 3 commits were to change from postfix + prefix to just a single transpiler. I think I made it fairly clear what it does and how it was made on the comment above it, but if more information is needed just tell me. 

I also made each CustomLanguage hold their tables instead of using a static dictionary (I don't know why I was before using a dict), so the TextTranslationPatches now only has patches.

**Comment by loco-choco** [comment]
grr when review

**Comment by MegaPiggy** [comment]
👴🏻

**Review by MegaPiggy — APPROVED** [review]

**Comment by MegaPiggy** [comment]
I don't have the power

**Review by xen-42 — APPROVED** [review]
ok why not!

**Comment by MegaPiggy** [comment]
only took 3 (almost 4) months

**Pull request merged**